### PR TITLE
[InstCombine] Simplify fractions when there is no overflow

### PR DESCRIPTION
**Pull request by AZero13**
Alive2 Proofs:
https://alive2.llvm.org/ce/z/7VTVWB
https://alive2.llvm.org/ce/z/-36zkQ
https://alive2.llvm.org/ce/z/epjaRg
https://alive2.llvm.org/ce/z/_jYLy4

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -1057,7 +1057,8 @@ static Value *foldIDivShl(BinaryOperator &I, InstCombiner::BuilderTy &Builder) {
 
     // (X * Y) s/ (X << Z) --> Y s/ (1 << Z)
     if (IsSigned && HasNSW && (Op0->hasOneUse() || Op1->hasOneUse())) {
-      Value *Shl = Builder.CreateShl(ConstantInt::get(Ty, 1), Z);
+      Value *Shl = Builder.CreateShl(ConstantInt::get(Ty, 1), Z, "", true,
+                                     HasNUW && HasNSW);
```
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1061]: This looks like it could be its own patch?

Why do you need to check HasNSW here? It was already checked on the `if` above.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1061]: True.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 2]
```diff
@@ -1273,13 +1314,30 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
...
-  // TODO: Handle sdiv.
   if (!IsSigned && Op1->hasOneUse() &&
       match(Op0, m_NUWShl(m_Value(X), m_Value(Z))) &&
       match(Op1, m_c_Mul(m_Specific(X), m_Value(Y))))
     if (cast<OverflowingBinaryOperator>(Op1)->hasNoUnsignedWrap()) {
       Instruction *NewDiv = BinaryOperator::CreateUDiv(
-          Builder.CreateShl(ConstantInt::get(Ty, 1), Z, "", /*NUW*/ true), Y);
+          Builder.CreateShl(
+              ConstantInt::get(Ty, 1), Z, "", /*NUW*/ true,
+              cast<OverflowingBinaryOperator>(Op1)->hasNoSignedWrap()),
```
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1324]: Changing the flags on an existing transform should be its own patch.

**Comment by AtariDreams** [comment]
Anyway, in addition to the Alive2 proof, in the event I messed up there, here is how if a multiplication followed by a division does not have an unsigned wrap, a reduced fraction does:

The smallest possible greatest common factor is 1. This means the fraction is reduced. For a fraction to not be reduce, it has to have a greatest common factor of at least 2.

A division of 2 is the same as a bit shift.

File: llvm/test/Transforms/InstCombine/div.ll [hunk 3]
```diff
@@ -380,7 +380,7 @@ define i32 @test26(i32 %a) {
 
 define i32 @test27(i32 %a) {
 ; CHECK-LABEL: @test27(
-; CHECK-NEXT:    [[DIV:%.*]] = shl nuw i32 [[A:%.*]], 1
+; CHECK-NEXT:    [[DIV:%.*]] = shl nuw nsw i32 [[A:%.*]], 1
```
Comment (topperc) [inline llvm/test/Transforms/InstCombine/div.ll:383]: Is this from a different change?

File: llvm/test/Transforms/InstCombine/div.ll [hunk 4]
```diff
@@ -1678,6 +1678,17 @@ define i32 @sdiv_sdiv_mul_nsw_exact_use(i32 %x, i32 %y, i32 %z) {
   ret i32 %r
 }
 
+define i32 @x_times_150_over_100(i32 %x) {
+; CHECK-LABEL: @x_times_150_over_100(
+; CHECK-NEXT:    [[TMP1:%.*]] = mul nuw nsw i32 [[X:%.*]], 3
+; CHECK-NEXT:    [[D1:%.*]] = lshr i32 [[TMP1]], 1
+; CHECK-NEXT:    ret i32 [[D1]]
+;
+  %m = mul nuw nsw i32 %x, 150
```
Comment (topperc) [inline llvm/test/Transforms/InstCombine/div.ll:1687]: Don't we need tests to cover all 3 cases here

```
          if (IsSigned && C1->isNegative() && C2->isNegative()) {
            NewC1 = C1->abs().udiv(GCD);
            NewC2 = C2->abs().udiv(GCD);
          } else if (IsSigned) {
            NewC1 = C1->sdiv(GCD);
            NewC2 = C2->sdiv(GCD);
          } else {
            NewC1 = C1->udiv(GCD);
            NewC2 = C2->udiv(GCD);
          }
```

**Comment by AtariDreams** [comment]
> > Anyway, in addition to the Alive2 proof, in the event I messed up there, here is how if a multiplication followed by a division does not have an unsigned wrap, a reduced fraction does:
> 
> The alive2 proofs are way to complicated. I don't think they really serve the purpose of avoiding bugs, as even though they verify, its not clear they don't have bugs in them.
> 
> I would suggest instead of implementing GCD, pass an additional argument and setup assumes s.t `GCD >= 1 && mulC % GCD == 0 && divC % GCD % 0`.

GCD being one means we will loop forever because every number that isn't zero has a common denominator of 1 with another

**Comment by nikic** [comment]
Could you please update the alive2 proofs in the way @goldsteinn suggested?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 5]
```diff
@@ -1176,6 +1176,45 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
...
         return Mul;
       }
+
+      // We can reduce expressions of things like * 150 / 100 to * 3 / 2
+      if (Op0->hasOneUse() && !C2->isZero() &&
+          !(IsSigned && C1->isMinSignedValue() && C2->isAllOnes())) {
+        assert(!C2->isMinSignedValue() &&
+               "This should have been folded away by InstSimplify");
+        APInt GCD = APIntOps::GreatestCommonDivisor(C1->abs(), C2->abs());
+        if (!GCD.isOne() && !GCD.isZero()) {
```
[the remark this answers is not in the document]
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1186]: > can GCD be zero?

No.

**Comment by AtariDreams** [comment]
> Could you please update the alive2 proofs in the way @goldsteinn suggested?

Done!

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 6]
```diff
@@ -1176,6 +1176,45 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
         Mul->setHasNoSignedWrap(OBO->hasNoSignedWrap());
         return Mul;
       }
+
+      // We can reduce expressions of things like * 150 / 100 to * 3 / 2
+      if (Op0->hasOneUse() && !C2->isZero() &&
+          !(IsSigned && C1->isMinSignedValue() && C2->isAllOnes())) {
+        assert(!C2->isMinSignedValue() &&
+               "This should have been folded away by InstSimplify");
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1184]: Not InstSimplify, InstCombine. It will get folded, but this shouldn't be an assert as we sometimes move around folds. Just check it in the above if statement.

**Comment by goldsteinn** [comment]
In your proofs please don't arbitrarily use `noundef`, and particularly there should never be a reason to `noundef` the return value. It hides bugs. In this case a bug with `sdiv INT_MIN, -1` that can result if `C1 == -C2` in the `sdiv` case:
https://alive2.llvm.org/ce/z/95jbW3

Its okay to `noundef` what can't be undef (mostly constants in your impl).

**Comment by goldsteinn** [comment]
Two further notes about your proofs:
1) This is general (regarding all your proofs across PRs). **Please label your proofs**. Either label the individual links, have a reasonably descriptive function name, or comment them (or all 3!). The way you do it now, where each proof is just a mystery link (and historically has been to completely incorrect/unrelated proofs) adds to reviewer burden. 
2) Make sure your proofs match your implementation, not your goal. I.e in this case you implement the `sdiv` case with `abs`. Your proofs should do that same.

But anyways, here are I *think* correct proofs: https://alive2.llvm.org/ce/z/1failT
The only missing check I see is that `NewC2 != -1` in the `sdiv` case. I think the `C2 != -1` is also gratuitous.
Please check over the proofs, if they make sense, please update your implementation to match their constraints. If you see issues, please update them and post back the corrected proofs (and maybe a correct impl).
